File: lua/test/async_order_detail_view_entity_test.lua
-- AsyncOrderDetailView entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("tangocard_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("AsyncOrderDetailViewEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:AsyncOrderDetailView(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = async_order_detail_view_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "async_order_detail_view." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local async_order_detail_view_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.async_order_detail_view")))
    local async_order_detail_view_ref01_data = nil
    if #async_order_detail_view_ref01_data_raw > 0 then
      async_order_detail_view_ref01_data = helpers.to_map(async_order_detail_view_ref01_data_raw[1][2])
    end

    -- UPDATE
    local async_order_detail_view_ref01_ent = client:AsyncOrderDetailView(nil)
    local async_order_detail_view_ref01_data_up0_up = {
      id = async_order_detail_view_ref01_data["id"],
      ["account_identifier"] = setup.idmap["account_identifier"],
      ["customer_identifier"] = setup.idmap["customer_identifier"],
    }

    local async_order_detail_view_ref01_markdef_up0_name = "accountIdentifier"
    local async_order_detail_view_ref01_markdef_up0_value = "Mark01-async_order_detail_view_ref01_" .. tostring(setup.now)
    async_order_detail_view_ref01_data_up0_up[async_order_detail_view_ref01_markdef_up0_name] = async_order_detail_view_ref01_markdef_up0_value

    local async_order_detail_view_ref01_resdata_up0_result, err = async_order_detail_view_ref01_ent:update(async_order_detail_view_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local async_order_detail_view_ref01_resdata_up0 = helpers.to_map(type(async_order_detail_view_ref01_resdata_up0_result) == 'table' and async_order_detail_view_ref01_resdata_up0_result.data_get and async_order_detail_view_ref01_resdata_up0_result:data_get() or async_order_detail_view_ref01_resdata_up0_result)
    assert.is_not_nil(async_order_detail_view_ref01_resdata_up0)
    assert.are.equal(async_order_detail_view_ref01_resdata_up0["id"], async_order_detail_view_ref01_data_up0_up["id"])
    assert.are.equal(async_order_detail_view_ref01_resdata_up0[async_order_detail_view_ref01_markdef_up0_name], async_order_detail_view_ref01_markdef_up0_value)

    -- LOAD
    local async_order_detail_view_ref01_match_dt0 = {
      id = async_order_detail_view_ref01_data["id"],
    }
    local async_order_detail_view_ref01_data_dt0_loaded, err = async_order_detail_view_ref01_ent:load(async_order_detail_view_ref01_match_dt0, nil)
    assert.is_nil(err)
    local async_order_detail_view_ref01_data_dt0_load_result = helpers.to_map(type(async_order_detail_view_ref01_data_dt0_loaded) == 'table' and async_order_detail_view_ref01_data_dt0_loaded.data_get and async_order_detail_view_ref01_data_dt0_loaded:data_get() or async_order_detail_view_ref01_data_dt0_loaded)
    assert.is_not_nil(async_order_detail_view_ref01_data_dt0_load_result)
    assert.are.equal(async_order_detail_view_ref01_data_dt0_load_result["id"], async_order_detail_view_ref01_data["id"])

  end)
end)

function async_order_detail_view_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/async_order_detail_view/AsyncOrderDetailViewTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read async_order_detail_view test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "async_order_detail_view01", "async_order_detail_view02", "async_order_detail_view03", "customer01", "customer02", "customer03", "account01", "account02", "account03", "accountentifier01", "customerentifier01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID"] = idmap,
    ["TANGOCARD_TEST_LIVE"] = "FALSE",
    ["TANGOCARD_TEST_EXPLAIN"] = "FALSE",
    ["TANGOCARD_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["account_identifier"] == nil then
    idmap_resolved["account_identifier"] = idmap_resolved["accountentifier01"]
  end
  if idmap_resolved["customer_identifier"] == nil then
    idmap_resolved["customer_identifier"] = idmap_resolved["customerentifier01"]
  end

  if env["TANGOCARD_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["TANGOCARD_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["TANGOCARD_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["TANGOCARD_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/async_order_detail_view_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/tangocard-sdk/go"
	"github.com/voxgig-sdk/tangocard-sdk/go/core"

	vs "github.com/voxgig-sdk/tangocard-sdk/go/utility/struct"
)

func TestAsyncOrderDetailViewEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.AsyncOrderDetailView(nil)
		if ent == nil {
			t.Fatal("expected non-nil AsyncOrderDetailViewEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := async_order_detail_viewBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "async_order_detail_view." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		asyncOrderDetailViewRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath(setup.data, "existing.async_order_detail_view")))
		var asyncOrderDetailViewRef01Data map[string]any
		if len(asyncOrderDetailViewRef01DataRaw) > 0 {
			asyncOrderDetailViewRef01Data = core.ToMapAny(asyncOrderDetailViewRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = asyncOrderDetailViewRef01Data

		// UPDATE
		asyncOrderDetailViewRef01Ent := client.AsyncOrderDetailView(nil)
		asyncOrderDetailViewRef01DataUp0Up := map[string]any{
			"id": asyncOrderDetailViewRef01Data["id"],
			"account_identifier": setup.idmap["account_identifier"],
			"customer_identifier": setup.idmap["customer_identifier"],
		}

		asyncOrderDetailViewRef01MarkdefUp0Name := "accountIdentifier"
		asyncOrderDetailViewRef01MarkdefUp0Value := fmt.Sprintf("Mark01-async_order_detail_view_ref01_%d", setup.now)
		asyncOrderDetailViewRef01DataUp0Up[asyncOrderDetailViewRef01MarkdefUp0Name] = asyncOrderDetailViewRef01MarkdefUp0Value

		asyncOrderDetailViewRef01ResdataUp0Result, err := asyncOrderDetailViewRef01Ent.Update(asyncOrderDetailViewRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		asyncOrderDetailViewRef01ResdataUp0 := core.ToMapAny(entityData(asyncOrderDetailViewRef01ResdataUp0Result))
		if asyncOrderDetailViewRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if asyncOrderDetailViewRef01ResdataUp0["id"] != asyncOrderDetailViewRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if asyncOrderDetailViewRef01ResdataUp0[asyncOrderDetailViewRef01MarkdefUp0Name] != asyncOrderDetailViewRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", asyncOrderDetailViewRef01MarkdefUp0Name, asyncOrderDetailViewRef01ResdataUp0[asyncOrderDetailViewRef01MarkdefUp0Name])
		}

		// LOAD
		asyncOrderDetailViewRef01MatchDt0 := map[string]any{
			"id": asyncOrderDetailViewRef01Data["id"],
		}
		asyncOrderDetailViewRef01DataDt0Loaded, err := asyncOrderDetailViewRef01Ent.Load(asyncOrderDetailViewRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		asyncOrderDetailViewRef01DataDt0LoadResult := core.ToMapAny(entityData(asyncOrderDetailViewRef01DataDt0Loaded))
		if asyncOrderDetailViewRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if asyncOrderDetailViewRef01DataDt0LoadResult["id"] != asyncOrderDetailViewRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func async_order_detail_viewBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "async_order_detail_view", "AsyncOrderDetailViewTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read async_order_detail_view test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse async_order_detail_view test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"async_order_detail_view01", "async_order_detail_view02", "async_order_detail_view03", "customer01", "customer02", "customer03", "account01", "account02", "account03", "accountentifier01", "customerentifier01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID": idmap,
		"TANGOCARD_TEST_LIVE":      "FALSE",
		"TANGOCARD_TEST_EXPLAIN":   "FALSE",
		"TANGOCARD_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["TANGOCARD_TEST_ASYNC_ORDER_DETAIL_VIEW_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add account_identifier alias for update test.
	if idmapResolved["account_identifier"] == nil {
		idmapResolved["account_identifier"] = idmapResolved["accountentifier01"]
	}
	// Add customer_identifier alias for update test.
	if idmapResolved["customer_identifier"] == nil {
		idmapResolved["customer_identifier"] = idmapResolved["customerentifier01"]
	}

	if env["TANGOCARD_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["TANGOCARD_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewTangocardSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TANGOCARD_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TANGOCARD_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
